tools/jit/gen_jit_dispatch.py

```diff
@@ -58,64 +63,126 @@ def is_jit_op(decl):
-        descriptor = '-'.join([decl['name'], str(num_inputs)] + attr_names)
-
-        # All scalar args need to be assigned, so they can be captured by a lambda
-        assignments = [ATTR_ASSIGNMENT.substitute(type=arg['simple_type'],
-                                                  type_cast=TYPE_CASTS.get(arg['simple_type'], arg['simple_type']),
-                                                  name=arg['name'],
-                                                  method=ATTR_METHOD_MAP[arg['simple_type']])
-                       for arg in scalar_args]
-
-        # Generate the actuall ATen call. This gets a bit tricky because of
-        # TensorList arguments, and functions that are only available as methods.
-        if 'namespace' in decl['method_of']:
-            if has_tensorlist:
-                if sum(map(is_tensor_arg, arguments)) != 1:
-                    # TODO: support this
-                    continue
-                args = ['TensorTemporaryList(inputs)' if is_tensor_arg(arg) else arg['name']
+        scalar_arg_idx = [i for i, arg in enumerate(arguments) if not is_tensor_arg(arg)]
+        tensor_arg_idx = [i for i, arg in enumerate(arguments) if is_tensor_arg(arg)]
+
+        # Right now, we generate dispatch methods that either take all non-tensor arguments
+        # as attributes, or don't use any attributes at all. In the future we might want to
+        # have something in the middle too (might be useful for e.g. constant propagation
+        # into attributes, as that would allow us to avoid reparsing tensors into scalar
+        # args at every invocation).
+        # NB: if there are no scalar args then both options on LHS are equivalent, so deduplicate them.
+        scalar_arg_idx_iter = ([], scalar_arg_idx) if scalar_arg_idx else ([],)
+        for pos_scalar_arg_idx in scalar_arg_idx_iter:
+            pos_scalar_args = [arguments[i] for i in pos_scalar_arg_idx]
+            kw_scalar_args = [arguments[i] for i in scalar_arg_idx
+                              if i not in pos_scalar_arg_idx]
+            # Descriptor is a unique identifier for a particular overload of an op
+            attr_names = sorted([arg['name'] for arg in kw_scalar_args])
+            num_inputs = len(tensor_arg_idx) + len(pos_scalar_arg_idx) if not has_tensorlist else '*'
+            descriptor = '-'.join([decl['name'], str(num_inputs)] + attr_names)
+
+            kw_assignments = [KW_ASSIGNMENT.substitute(type_cast=TYPE_CASTS.get(arg['simple_type'], arg['simple_type']),
+                                                       name=arg['name'],
+                                                       method=ATTR_METHOD_MAP[arg['simple_type']])
+                              for arg in kw_scalar_args]
+            # TODO: if has_tensor_list then we need to count form the end!
+            pos_scalar_arg_offsets = [i + len([None for t_idx in tensor_arg_idx if t_idx < s_idx])
```

Comment:
`[None for t_idx in tensor_arg_idx if t_idx < s_idx]` <--  This heavy use of array indexing is confusing. I can't really follow what is going on here. Can you find a clearer way of categorizing each input? I think a loop over all arguments, classifying them as tensor, position_scalar, or kwarg and adding entries into pos_assigns, kw_assignments, etc. would be easier to follow.